File: bindings/python/triepack/decoder.py
# Copyright (c) 2026 M. A. Chatterjee, BSD-2-Clause.
"""Trie decoder — port of src/core/decoder.c (via JavaScript decoder.js)."""

from .bitstream import BitReader
from .crc32 import crc32
from .values import decode_value
from .varint import read_var_uint

# Format constants
TP_MAGIC = bytes([0x54, 0x52, 0x50, 0x00])
TP_HEADER_SIZE = 32
TP_FLAG_HAS_VALUES = 1
NUM_CONTROL_CODES = 6

# Control code indices
CTRL_END = 0
CTRL_END_VAL = 1
CTRL_SKIP = 2
CTRL_BRANCH = 5


def decode(buffer):
    """Decode a .trp binary buffer into a dict. Returns dict."""
    if not isinstance(buffer, (bytes, bytearray)):
        raise TypeError("decode expects bytes or bytearray")
    if len(buffer) < TP_HEADER_SIZE + 4:
        raise ValueError("Data too short for .trp format")

    buf = bytes(buffer)

    # Validate magic
    for i in range(4):
        if buf[i] != TP_MAGIC[i]:
            raise ValueError("Invalid magic bytes — not a .trp file")

    # Version
    version_major = buf[4]
    if version_major != 1:
        raise ValueError(f"Unsupported format version: {version_major}")

    # Flags
    flags = (buf[6] << 8) | buf[7]
    has_values = (flags & TP_FLAG_HAS_VALUES) != 0

    # num_keys
    num_keys = (buf[8] << 24) | (buf[9] << 16) | (buf[10] << 8) | buf[11]

    # Offsets
    trie_data_offset = (buf[12] << 24) | (buf[13] << 16) | (buf[14] << 8) | buf[15]
    value_store_offset = (buf[16] << 24) | (buf[17] << 16) | (buf[18] << 8) | buf[19]

    # CRC verification
    crc_data_len = len(buf) - 4
    expected_crc = (
        (buf[crc_data_len] << 24)
        | (buf[crc_data_len + 1] << 16)
        | (buf[crc_data_len + 2] << 8)
        | buf[crc_data_len + 3]
    )
    actual_crc = crc32(buf[:crc_data_len])
    if actual_crc != expected_crc:
        raise ValueError("CRC-32 integrity check failed")

    if num_keys == 0:
        return {}

    # Parse trie config
    reader = BitReader(buf)
    data_start = TP_HEADER_SIZE * 8

    reader.seek(data_start)
    bps = reader.read_bits(4)
    # Symbol codes index 256-entry maps, so a symbol may not be wider than a
    # byte; 0 would make the trie unreadable.
    if bps < 1 or bps > 8:
        raise ValueError(f"Invalid trie config: bits_per_symbol must be 1..8, got {bps}")
    symbol_count = reader.read_bits(8)
    # Must leave room for the control codes and fit in bits_per_symbol.
    if symbol_count < NUM_CONTROL_CODES or symbol_count > (1 << bps):
        raise ValueError(f"Invalid trie config: symbol_count out of range: {symbol_count}")

    # Read control codes
    ctrl_codes = [0] * NUM_CONTROL_CODES
    code_is_ctrl = [False] * 256
    for c in range(NUM_CONTROL_CODES):
        ctrl_codes[c] = reader.read_bits(bps)
        if ctrl_codes[c] < 256:
            code_is_ctrl[ctrl_codes[c]] = True

    # Read symbol table
    reverse_map = [0] * 256
    for cd in range(NUM_CONTROL_CODES, symbol_count):
        cp = read_var_uint(reader)
        if cd < 256 and cp < 256:
            reverse_map[cd] = cp

    trie_start = data_start + trie_data_offset
    value_start = data_start + value_store_offset
    # The trie occupies [trie_start, value_start); the value store (when
    # present), the byte padding and the CRC follow it.
    trie_end = value_start

    # DFS iteration
    result = {}
    key_stack = []

    # Every subtree knows where it ends: a child with a SKIP ends at
    # child_start + skip_dist, the last child ends where its parent does, and
    # the root ends at trie_end. `end` is the only authority on whether more
    # symbols belong to this subtree -- the bits that happen to follow are
    # not, because past the last terminal they are padding and CRC.
    def dfs_walk(r, end):
        while True:
            if r.position >= end:
                return
            sym = r.read_bits(bps)

            if sym == ctrl_codes[CTRL_END]:
                key_str = bytes(key_stack).decode("utf-8")
                result[key_str] = None
                walk_branch_if_present(r, end)
                return

            if sym == ctrl_codes[CTRL_END_VAL]:
                read_var_uint(r)  # value index
                key_str = bytes(key_stack).decode("utf-8")
                result[key_str] = None
                walk_branch_if_present(r, end)
                return

            if sym == ctrl_codes[CTRL_BRANCH]:
                child_count = read_var_uint(r)
                walk_branch(r, child_count, end)
                return

            # Regular symbol
            byte_val = reverse_map[sym] if sym < 256 else 0
            key_stack.append(byte_val)

    # A terminal is followed by a BRANCH exactly when the subtree has not
    # reached its end -- keys that share this terminal as a prefix.
    def walk_branch_if_present(r, end):
        if r.position >= end:
            return
        sym = r.read_bits(bps)
        if sym != ctrl_codes[CTRL_BRANCH]:
            raise ValueError("Malformed trie: expected BRANCH after terminal")
        child_count = read_var_uint(r)
        walk_branch(r, child_count, end)

    def walk_branch(r, child_count, end):
        saved_key_len = len(key_stack)
        for ci in range(child_count):
            has_skip = ci < child_count - 1
            skip_dist = 0

            if has_skip:
                r.read_bits(bps)  # SKIP control code
                skip_dist = read_var_uint(r)

            child_start_pos = r.position
            del key_stack[saved_key_len:]
            dfs_walk(r, child_start_pos + skip_dist if has_skip else end)

            if has_skip:
                r.seek(child_start_pos + skip_dist)

        del key_stack[saved_key_len:]

    # Walk the trie
    reader.seek(trie_start)
    if num_keys > 0:
        dfs_walk(reader, trie_end)

    # Decode values
    if has_values:
        reader.seek(value_start)
        sorted_keys = sorted(result.keys(), key=lambda k: k.encode("utf-8"))
        for key in sorted_keys:
            result[key] = decode_value(reader)

    return result

File: bindings/go/decoder.go
// Copyright (c) 2026 M. A. Chatterjee, BSD-2-Clause.

package triepack

import (
	"fmt"
	"sort"
)

// decodeData decodes a .trp binary buffer into a map.
func decodeData(buffer []byte) (map[string]interface{}, error) {
	if len(buffer) < tpHeaderSize+4 {
		return nil, fmt.Errorf("data too short for .trp format")
	}

	buf := buffer

	// Validate magic
	for i := 0; i < 4; i++ {
		if buf[i] != tpMagic[i] {
			return nil, fmt.Errorf("invalid magic bytes - not a .trp file")
		}
	}

	// Version
	versionMajor := buf[4]
	if versionMajor != 1 {
		return nil, fmt.Errorf("unsupported format version: %d", versionMajor)
	}

	// Flags
	flags := uint16(buf[6])<<8 | uint16(buf[7])
	hasValues := (flags & tpFlagHasValues) != 0

	// num_keys
	numKeys := int(uint32(buf[8])<<24 | uint32(buf[9])<<16 | uint32(buf[10])<<8 | uint32(buf[11]))

	// Offsets
	trieDataOffset := int(uint32(buf[12])<<24 | uint32(buf[13])<<16 | uint32(buf[14])<<8 | uint32(buf[15]))
	valueStoreOffset := int(uint32(buf[16])<<24 | uint32(buf[17])<<16 | uint32(buf[18])<<8 | uint32(buf[19]))

	// CRC verification
	crcDataLen := len(buf) - 4
	expectedCRC := uint32(buf[crcDataLen])<<24 | uint32(buf[crcDataLen+1])<<16 | uint32(buf[crcDataLen+2])<<8 | uint32(buf[crcDataLen+3])
	actualCRC := computeCRC32(buf[:crcDataLen])
	if actualCRC != expectedCRC {
		return nil, fmt.Errorf("CRC-32 integrity check failed")
	}

	if numKeys == 0 {
		return make(map[string]interface{}), nil
	}

	// Parse trie config
	reader := NewBitReader(buf)
	dataStart := tpHeaderSize * 8

	reader.Seek(dataStart)
	bpsBits, err := reader.ReadBits(4)
	if err != nil {
		return nil, err
	}
	bps := int(bpsBits)

	symCountBits, err := reader.ReadBits(8)
	if err != nil {
		return nil, err
	}
	symbolCount := int(symCountBits)

	// Read control codes
	var ctrlCodes [numControlCodes]int
	for c := 0; c < numControlCodes; c++ {
		v, err := reader.ReadBits(bps)
		if err != nil {
			return nil, err
		}
		ctrlCodes[c] = int(v)
	}

	// Read symbol table
	var reverseMap [256]int
	for cd := numControlCodes; cd < symbolCount; cd++ {
		cp, err := readVarUint(reader)
		if err != nil {
			return nil, err
		}
		if cd < 256 && cp < 256 {
			reverseMap[cd] = int(cp)
		}
	}

	trieStart := dataStart + trieDataOffset
	valueStart := dataStart + valueStoreOffset
	// The trie occupies [trieStart, valueStart); the value store (when
	// present), the byte padding and the CRC follow it.
	trieEnd := valueStart

	// DFS iteration
	result := make(map[string]interface{})
	keyStack := make([]byte, 0, 64)

	var dfsWalk func(r *BitReader, end int) error
	var walkBranchIfPresent func(r *BitReader, end int) error
	var walkBranch func(r *BitReader, childCount int, end int) error

	// Every subtree knows where it ends: a child with a SKIP ends at
	// childStart + skipDist, the last child ends where its parent does, and
	// the root ends at trieEnd. end is the only authority on whether more
	// symbols belong to this subtree - the bits that happen to follow are
	// not, because past the last terminal they are padding and CRC.
	dfsWalk = func(r *BitReader, end int) error {
		for {
			if r.Position() >= end {
				return nil
			}
			sym, err := r.ReadBits(bps)
			if err != nil {
				return err
			}
			s := int(sym)

			if s == ctrlCodes[ctrlEnd] {
				keyStr := string(keyStack)
				result[keyStr] = nil
				return walkBranchIfPresent(r, end)
			}

			if s == ctrlCodes[ctrlEndVal] {
				_, err := readVarUint(r) // value index
				if err != nil {
					return err
				}
				keyStr := string(keyStack)
				result[keyStr] = nil
				return walkBranchIfPresent(r, end)
			}

			if s == ctrlCodes[ctrlBranch] {
				cc, err := readVarUint(r)
				if err != nil {
					return err
				}
				return walkBranch(r, int(cc), end)
			}

			// Regular symbol
			byteVal := byte(0)
			if s < 256 {
				byteVal = byte(reverseMap[s])
			}
			keyStack = append(keyStack, byteVal)
		}
	}

	// A terminal is followed by a BRANCH exactly when the subtree has not
	// reached its end - keys that share this terminal as a prefix.
	walkBranchIfPresent = func(r *BitReader, end int) error {
		if r.Position() >= end {
			return nil
		}
		sym, err := r.ReadBits(bps)
		if err != nil {
			return err
		}
		if int(sym) != ctrlCodes[ctrlBranch] {
			return fmt.Errorf("malformed trie: expected BRANCH after terminal")
		}
		cc, err := readVarUint(r)
		if err != nil {
			return err
		}
		return walkBranch(r, int(cc), end)
	}

	walkBranch = func(r *BitReader, childCount int, end int) error {
		savedKeyLen := len(keyStack)
		for ci := 0; ci < childCount; ci++ {
			hasSkip := ci < childCount-1
			skipDist := 0

			if hasSkip {
				_, err := r.ReadBits(bps) // SKIP control code
				if err != nil {
					return err
				}
				sd, err := readVarUint(r)
				if err != nil {
					return err
				}
				skipDist = int(sd)
			}

			childStartPos := r.Position()
			keyStack = keyStack[:savedKeyLen]
			childEnd := end
			if hasSkip {
				childEnd = childStartPos + skipDist
			}
			if err := dfsWalk(r, childEnd); err != nil {
				return err
			}

			if hasSkip {
				r.Seek(childStartPos + skipDist)
			}
		}
		keyStack = keyStack[:savedKeyLen]
		return nil
	}

	// Walk the trie
	reader.Seek(trieStart)
	if numKeys > 0 {
		if err := dfsWalk(reader, trieEnd); err != nil {
			return nil, err
		}
	}

	// Decode values
	if hasValues {
		reader.Seek(valueStart)
		sortedKeys := make([]string, 0, len(result))
		for k := range result {
			sortedKeys = append(sortedKeys, k)
		}
		sort.Slice(sortedKeys, func(i, j int) bool {
			return sortedKeys[i] < sortedKeys[j]
		})
		for _, key := range sortedKeys {
			v, err := decodeValue(reader)
			if err != nil {
				return nil, err
			}
			result[key] = v
		}
	}

	return result, nil
}
